src/util.rs

```diff
@@ -39,143 +43,216 @@ pub(crate) fn parse_from_string(input: &str) -> bril_rs::Program {
+pub fn visualize(program_path: PathBuf, output_dir: PathBuf) -> io::Result<()> {
+    use std::fs::File;
+    use std::io::Write;
+
+    // make the directory if it doesn't exist
+    if !output_dir.exists() {
+        std::fs::create_dir_all(&output_dir)?;
+    }
+
+    let all_configs = Run::all_configurations_for(program_path);
+
+    let results = all_configs.iter().map(|run| (run, run.run()));
+
+    for (run, result) in results {
+        let mut output_path = output_dir.clone();
+        assert!(output_path.is_dir());
+        output_path.push(format!(
+            "{}{}",
+            run.name(),
+            result.visualization_file_extension
+        ));
+        let mut file = File::create(output_path)?;
+        if let Some(interpreted) = result.result_interpreted {
+            file.write_all(interpreted.as_bytes())?;
+        } else {
+            file.write_all(result.visualization.as_bytes())?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Invokes some program with the given arguments, piping the given input to the program.
+/// Returns an error if the program returns a non-zero exit code.
+/// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
+pub fn run_cmd_line<S1, S2, I>(program: S1, args: I, input: &str) -> std::io::Result<String>
+where
+    S1: AsRef<OsStr>,
+    S2: AsRef<OsStr>,
+    I: IntoIterator<Item = S2>,
+{
+    use std::io::Write;
+    use std::process::{Command, Stdio};
+    let mut child = Command::new(program)
+        .args(args)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let stdin = child.stdin.as_mut().expect("Failed to open stdin");
+    write!(stdin, "{}", input)?;
+
+    let output = child.wait_with_output()?;
+    match output.status.code() {
+        Some(0) => Ok(String::from_utf8(output.stdout).map_err(|e| {
+            std::io::Error::new(std::io::ErrorKind::Other, format!("utf8 error: {}", e))
+        })?),
+        Some(e) => Err(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            format!("program returned error code {}", e),
+        )),
+        None => Err(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "program was killed by a signal",
+        )),
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum RunType {
+    StructuredConversion,
+    RvsdgConversion,
+    NaiiveOptimization,
+}
+
+impl Debug for RunType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl FromStr for RunType {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "structured" => Ok(RunType::StructuredConversion),
+            "rvsdg" => Ok(RunType::RvsdgConversion),
+            "naiive" => Ok(RunType::NaiiveOptimization),
+            _ => Err(format!("Unknown run type: {}", s)),
+        }
+    }
 }
 
-impl DebugVisualizations {
```

Comment:
It's not clear to me how easy it is to add ad-hoc calls to this code now. One thing I use DebugVisualizations for is just adding a call to it in a test that I'm debugging. Can we still do that here?

Reply:
I'll change the `visualize` function to take a bril program instead of a file path. Does that recover the functionality?

Reply:
I've refactored the code so that you can either provide a bril file or the actual bril code (along with the name for the code and the arguments)